sdk/servicebus/Microsoft.Azure.WebJobs.Extensions.ServiceBus/src/Listeners/ServiceBusScaleMonitor.cs

```diff
@@ -144,7 +144,7 @@ internal static ServiceBusTriggerMetrics CreateTriggerMetrics(ServiceBusReceived
             long totalNewMessageCount = 0;
             TimeSpan queueTime = TimeSpan.Zero;
 
-            if (message != null)
+            if (message != null && MessageWasNotScheduled(message))
```

Comment:
It won't be the case that there are deferred messages at the top of the queue with non-deferred messages after those somehow? I.e. if peek returns a deferred message, it will always be the case that a receive won't receive anything because there aren't any active/actionable messages?

It's somewhat confusing that [PeekMessageAsync](https://github.com/Azure/azure-sdk-for-net/blob/53c46a70140bcdc3041968fbbe0b1e30c282c1be/sdk/servicebus/Azure.Messaging.ServiceBus/src/Receiver/ServiceBusReceiver.cs#L372) says it returns **active** messages, but in the fine print says it returns deferred as well.

Reply:
Also, I'll point out that changes here will only affect apps configured to use [Runtime Scale Monitoring](https://docs.microsoft.com/en-us/azure/azure-functions/functions-networking-options#premium-plan-with-virtual-network-triggers). We'd need to make the same change in ScaleController as well for regular Consumption SKUs.

Reply:
I was only able to fix the code that is publicly accessible correct. If the other parts are available and you can point me to it I'm happy to contribute the changes there too but I seem to understand they are internal

Reply:
The peeked message is used to determine the Queue Time. This code only ever picks one message per call and assumes further calls will return more valid messages so I preserved that since scheduled messages are not valid messages. When manage rights are present the admin clients are used to get a count of the actual active messages.

I was thinking first about using PeekMessagesAsync (notice the plural) and skip until a valid message is found but then it would be necessary to come up with a reasonable Nummer of messages to peek and decide whether multiple peeks within the same metrics retrieval would be OK which has an impact on the runtime aspects. Given that this is my first contribution to this code base and I lack deep understanding I figured I go with the most straightforward approach and then adjust as needed.

Reply:
Forgot to mention a detail. Peek remembers the last peeked sequence number and therefore subsequent peeks move through the queue

This is the crucial behaviour this code was and is relying upon

Reply:
Yes, the consumption scale code is internal - I was just pointing this out for everyone looking at this, that we'll want to make the same change to SC. The answer I'm trying to get regarding the peek operation is if we can always rely on there being no actionable messages if the first message peeked is deferred. I.e. does the server order these somehow so that if there are non-deferred actionable messages, those will be returned before any deferred messages. Otherwise, this change wouldn't work as intended.

Reply:
I have a small harness I can use to verify that maybe next week or maybe Josh can chime in in case he knows the behaviour of the method. Unfortunately the documentation doesn't make it very clear

Reply:
`PeekMessageAsync` returns them in the order of the `SequenceNumber`. So that means we have to do multiple peeks until it returns a valid message or nothing anymore which means the queue is empty. Any preference on the way this should be done? Always peek a few messages or just one first and when it is a scheduled one start fetching multiple? Or always one at the time at the expense of having more latency per call but prevent over peeking?

Reply:
I'm not sure I understand the implications of this returning 1 vs 0 for the no-manage-claims case, so it is hard to say what the benefit would be of peeking more messages here. Is it very impactful to the runtime scaling behavior when there is 1 new message vs 0 new messages, @mathewc? In any case, I think we'd want to limit the peeking to the case where manage claims are not there.

Reply:
Any hints or ideas? Otherwise I think I'm a bit stuck here and can't proceed

Reply:
I think the issue is orthogonal to whether there are manage claims or not. If there's a scheduled/deferred message at the head of the queue, queue time will steadily increase over N samples, triggering a scale out vote. Ideally we'd have a way to ignore any such messages and only base the queue time check on active messages.

Reply:
>  and only base the queue time check on active messages.
But this is what you get with the Administration client when manage claims are available

Reply:
The admin client gives you message count which is used for a few of the scale heuristics. Queue time is used in a separate heuristic. E.g. if you have managed claims and know the queue count is 100 and steady that won't trigger the count based heuristics to scale out. You also want to take into account whether any progress is being made on the messages - if the queue time is monotonically increasing that's another indicator that the current workers aren't sufficient.

Reply:
Ah I see - I missed the queue time part of it.

Reply:
I just tested this out and I confirmed that the peek order is just based on the order of when the messages are initially enqueued (or scheduled). So if you send a message and then defer it, and then schedule a message, and then send a regular message, the peek order would be Deferred, Scheduled, Active.

Reply:
I've pushed a simple approach that peeks multiple messages and skips until an active message was found or nothing is returned. This is to get the discussion going. Maybe it makes sense to:

- Peek first with count 1 and then gradually increase over multiple peeks towards a defined max. What would that be?
- Add some kind of exit criteria based on what? Max Number of messages we want to fetch per cycle or time that has passed between fetches?
- Ideally we would also want a cancellation token that tells us to stop peeking when the host shuts down, but I couldn't find something available at first sight. Any ideas?

Reply:
@JoshLove-msft @mathewc any inputs from your side regarding the above and the current implementation? I'm happy to spend a bit more effort on this as long as I know I'm heading in a direction that is aligned with you guys

Reply:
I don't think we can reliably choose an N for the number of messages to peek, so this would really be a best effort. One thing we could do is store off the last peeked sequence number if all messages that are returned are scheduled or deferred. This way, we have a better chance of getting an active message in the next GetMetrics call.

Reply:
Addressed